### ISSUE TITLE
_document.js does not work for React.JS v18

### ISSUE THREAD

**Issue by Harry-Yep** [vercel/next.js#35711]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:36:53 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T8101
    Binaries:
      Node: 17.8.0
      npm: 8.5.5
      Yarn: 1.22.18
      pnpm: N/A
    Relevant packages:
      next: 12.1.2
      react: 18.0.0
      react-dom: 18.0.0
```

### Describe the Bug

`_document.js` does not work for React.JS `v18`

### Expected Behavior

`_document.js` and components such as `next/head` should work for React.JS v18

### To Reproduce

create a basic next.js project and create a `_document.js` file in `/pages`.

**Comment by SukkaW** [comment]
Have you tried `react@18` with `next@12.1.1`?

I have run into a similar issue, and downgrading to `next@12.1.1` the issue is gone. I am still comparing the changes to see which change might cause the issue.

**Comment by Harry-Yep** [comment]
Hi @SukkaW

Yes. I have tried to downgrade to `next@12.1.1` before, it works, but I don't think this is a long-term solution.

I recognize that the changes between `12.1.1` and `12.1.2-canary.0` caused _document.js to malfunction, as you found.

Looking forward to a further answer & solution to this issue.

**Comment by timneutkens** [comment]
We're working on a fix for this issue in #35736. Will be out on stable today 🙏
